### DISCUSSION TITLE
PyMC api for construction and inference with Bayesian Networks

### DISCUSSION THREAD

**Discussion by arainboldt** [pymc-devs/pymc3#6625] [Ideas]
Hi all,

In response to [this](https://discourse.pymc.io/t/bayes-nets-belief-networks-and-pymc/5150/8) thread from @drbenvincent on the topic of bayesian networks with PyMC, and in followup to [this](https://github.com/pymc-labs/CausalPy/issues/173) thread, I wanted to see if there was any general interest in developing a PyMC api for defining and inferring with bayesian networks. 

What alternatives exist and why not use them?

While there are a handful of libraries that cover a good deal of the needs for working with discrete bayesnets, e.g. [pommegranate](https://pomegranate.readthedocs.io/en/latest/), [bnlearn](https://erdogant.github.io/bnlearn/pages/html/index.html), the former doesn't utilize mcmc for parameter learning, and the latter doesn't offer the flexibility and power that the PyMC ecosystem does. So far as I can tell at least. 

It'd be nice to know what other people might be interested in using a bayesian networks defined in PyMC for. I'll share my currently desired use case.

My current use case:

My current project is using discrete data that has some very poorly represented portions of the probability space. Consequently, the straightforward conditional-probability-table method that's usually employed with discrete BNs will leave a lot of zero-probability zones in our model that shouldn't be completely zero. Additionally, our data has a natural hierarchy. Consequently, we'd like to utilize the natural structure in the data to fill in the gaps, so to speak. bayesian hierachical modelling is a natural solution to this issue, but I'm not aware of how this could be done with existing BN tools outside of PyMC. Please share any suggestions.

In addition to the above reason, PyMC is just a pleasure to work with and Bayes Nets seem like a natural extension of its application. 

Love to hear what others' thoughts are and if there's any interest in contributing to developing a BN api for PyMC.

Thanks

**Comment by kpapdac** [comment]
What a great idea! 
I think what both bnlearn and pommegranate are missing is the ability to compute hybrid bayesian networks (mix of continuous and discrete variables). 
Marco Scutari in chapter 5 of his [book](https://www.taylorfrancis.com/books/mono/10.1201/9780429347436/bayesian-networks-marco-scutari-jean-baptiste-denis), summarizes nicely the pros and cons of the different software packages in BNs back then. BUGS/stan/pymc can address the computational challenge there by solving hybrid networks using approximate inference (see ch3). 

For pymc and [pyro](https://forum.pyro.ai/t/is-it-possible-to-construct-a-bayesian-network-with-pyro/421/2) using either MCMC or SVI it would be easy to configure such networks both at a top and hierarchical level. I'd be very interested - would the BN api potentially include structure learning too?

**Reply by arainboldt** [reply]
Good point about mixed models. This could be very intuitive in pymc. Structure learning would also be really cool to have!

**Reply by drbenvincent** [reply]
I don't know much about structure learning, but there are packages which do this. So I think there would have to be a clear value add compared to those. 

My suspicion would be that this would be more suitable for CausalPy (maybe), but not PyMC.

**Reply by arainboldt** [reply]
Good point. One added value is having all the tools for a workflow in one place. Structure learning, however, is a rather diverse field in itself, could be a bit of project to implement all the current algorithms in pymc/aesara.

**Comment by ricardoV94** [comment]
@arainboldt do you have some ideas of what such API could look like and what would it be doing behind the scenes?

**Reply by arainboldt** [reply]
Coming at this from a graph perspective facilitates my thinking, so I'll express myself in those terms. 

There are at least 3 levels of abstraction that could be better facilitated by a bayesnet api. First a DAG, which is a collection of relations between nodes. Second a node which is a variable, observed or not, that also represents a set of relations with neighbors. Third, an edge which is a relation between nodes.

What I'm thinking is essentially an API that would facilitate composing pymc models at each of these levels of abstraction. 

First, at a high level I was thinking the API could center around an object where one can define and infer from the DAG. So it would either:

1. take a dag and data and infer parameters. (this assumes a non-parametric approach to conditional distributions, which is intuitive for discrete variables, but perhaps less so for continuous)
e.g. something like `pm.BayesNet(nx.DiGraph(edges), data)`

3. or, allow one to build a dag progressively, similar to what is currently done with the context manager method for building models. 

Second, a way to define a Node and connect it to it's parents and children. While a node may simply be a `pm.Distribution`, it could also be an entire model. 

```
cloudy = pm.Node('Cloudy, observe=data['cloudy'])
season = pm.Node('Season, observe=data['season'])
rain = pm.Node(name='rain', observe=data['rain'], parents=[cloudy, season])
wet = pm.Node('Wet', observe=data['wet'], parents=[rain])
```

Finally, a way to define dependencies. In my experience this is either done with conditional probability tables (CPT) in the discrete case or alternately with a model or function that defines the relationship between parents and child. The `pm.Model` object already serves this functional approach, but the CPT approach is cumbersome in pymc. In the above example, the CPT for P(rain | cloudy, season) could be inferred directly from the data, as is done in most bayesnet packages, or something more explicit and creative could be passed to a specific parameter of the distribution underlying the given pm.Node object. 

I feel there is a way to better facilitate defining and inferring parameters of conditional distributions from many observed variables. [here](https://stats.stackexchange.com/questions/146814/bayes-net-parameter-learning-in-pymc), [here](https://discourse.pymc.io/t/intercausal-reasoning-in-bayesian-networks/2419), [here](https://discourse.pymc.io/t/observed-data-in-bayesian-networks/5513/12) and elsewhere are examples of bayesnets built in pymc, but the manner in which they're defined isn't easy to use or intuitive for me. 

So as far as the look and use of the API, I would like to see something that served these needs. Additionally, as @kpapdac mentioned, structure learning would also be a nice addition.

As for the backend, I don't expect that anything is required that isn't already present in pymc. For me it's really about making it easier to define a model from the perspective of a bayesian network that entails observing a joint distribution of 15-20 variables, define their connections and relations and then infer unknown parameters, perform causal inference, etc.

**Reply by ricardoV94** [reply]
If there are no immediate backend limitations, I think a Proof-of-concept API would be quite welcomed at [pymc-experimental](https://github.com/pymc-devs/pymc-experimental)

**Reply by arainboldt** [reply]
Sounds good. What's the process for submitting POCs? I didn't find a "how to contribute" process described anywhere. Is it preferred to do this via a PR or notebook?

**Reply by ricardoV94** [reply]
A pull request is ideal. A PR can just contain a notebook during the early phases or already example source code that would be integrated into the package (or both).

Whatever feels more natural for you.

We have some contributing guides here: https://www.pymc.io/projects/docs/en/latest/contributing/index.html

**Reply by arainboldt** [reply]
Ah ok. Awesome, thanks!
